### PR TITLE
[Validator] add groups support to the Valid constraint

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #3622, #17622
| License       | MIT
| Doc PR        | TODO

### CODE REVIEW

File: src/Symfony/Component/Validator/Constraints/ValidValidator.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ */
+class ValidValidator extends ConstraintValidator
+{
+    public function validate($value, Constraint $constraint)
+    {
+        if (!$constraint instanceof Valid) {
+            throw new UnexpectedTypeException($constraint, __NAMESPACE__.'\Valid');
+        }
+
+        $violations = $this->context->getValidator()->validate($value, null, array($this->context->getGroup()));
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/ValidValidator.php]: I am not completely sure if handling the traversal really should be done here or if we should better add some special handling for the `Valid` constraint to the `RecursiveContextualValidator` instead. Any opinions are warmly welcome.

**Comment by kgilden** [comment]
@xabbuh awesome! I confirmed it works as would validation groups work for every other constraint, i.e. in `validation.xml` you add groups to ´Valid` and now pass the `validation_groups` option when creating the form. `Valid` will only be applied if any of the groups in `validation_groups` equals those specified in `validation.xml`.

However, I do think there is one super useful use case not covered. Consider an `Profile` object having `profilePicture` and `coverPhoto` properties - both of type `Image`, which internally has a `file` property of type `File`. Let's say we have different maximum file sizes for both of them, so we add 2 `File` constraints to the `Image` object. And both properties should be valid at all times for `Profile`.

```xml
<class name="Profile">
  <property name="profilePicture">
    <constraint name="Valid" />
  </property>
  <property name="coverPhoto">
    <constraint name="Valid" />
  </property>
</class>

<class name="Image">
  <property name="file">
    <constraint name="File">
      <option name="maxSize">100k</option>
      <option name="groups">
        <value>profile-picture</value>
      </option>
    </constraint>
    <constraint name="File">
      <option name="maxSize">1000k</option>
      <option name="groups">
        <value>cover-photo</value>
      </option>
    </constraint>
  </property>
</class>
```

In `ProfileType` we'd have something like this probably.

```php
$builder->add('profilePicture', ImageType::class, ['validation_groups' => ['profile-picture]]);
$builder->add('coverPhoto', ImageType::class, ['validation_groups' => ['cover-photo']);
```

But try as you might, it seems to me that validation for these `File` objects does not kick in.

```
$form = $factory->create(ProfileType::class);
$form->handleRequest($request);
$form->isValid(); // returns true
```

**Comment by xabbuh** [comment]
@kgilden This seems logical to me as the root form itself is validated in the `Default` group. Configuration `validation_groups` for the root form to be, for example, `['Default', 'profile-picture', 'cover-photo']` should do the trick.

**Comment by kgilden** [comment]
@xabbuh ahh yes, sorry for the confusion. I should note that this patch is *still* required to make my example above work. The same groups must also be added to `Valid` constraints and then it works as expected.

**Comment by xabbuh** [comment]
@kgilden You are right. But IMO this behaviour is straight-forward with how validation groups in Symfony work in general. Do you disagree with that?

**Comment by kgilden** [comment]
@xabbuh it's straight-forward if for example the `File` constraint is applied to two distinct properties on the root object, i.e.

```
<class name="Profile">
  <property name="profilePicture">
    <constraint name="File">
        <!-- distinct options -->
    </constraint>
  </property>
  <property name="coverPhoto">
    <constraint name="File">
        <!-- distinct options -->
    </constraint>
  </property>
</class>
```

What I think is currently not possible is when you want to validate the same objects deeper. After this PR it's possible to do as follows.

```xml
<class name="Profile">
  <property name="profilePicture">
    <constraint name="Valid">
      <option name="groups">
        <value>profile-picture</value>
      </option>
    </constraint>
  </property>
  <property name="coverPhoto">
    <constraint name="Valid">
      <option name="groups">
        <value>cover-photo</value>
      </option>
    </constraint>
  </property>
</class>

<class name="Image">
  <property name="file">
    <constraint name="File">
      <option name="maxSize">100k</option>
      <option name="groups">
        <value>profile-picture</value>
      </option>
    </constraint>
    <constraint name="File">
      <option name="maxSize">1000k</option>
      <option name="groups">
        <value>cover-photo</value>
      </option>
    </constraint>
  </property>
</class>
```

```php
class ProfileType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder->add('profilePicture', ImageType::class);
        $builder->add('coverPhoto', ImageType::class);
    }

    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefaults(['validation_groups' => ['Default', 'cover-photo', 'profile-picture']]);
    }
}
```

Correct me if I'm wrong, but I don't think there was a way to nicely do it prior to this PR.

**Comment by xabbuh** [comment]
Indeed, that's not possible right now.

**Comment by pourquoi** [comment]
+1 for this feature. The use case described by kgilden happens regularly for me (deep cascade validation groups)

**Comment by fabpot** [comment]
@xabbuh Can you add a note in the CHANGELOG before I can merge?

**Comment by xabbuh** [comment]
@fabpot done, but are we sure that this is the right way to implement it?

**Comment by xabbuh** [comment]
ping @symfony/deciders

**Review by fabpot — APPROVED** [review]

File: src/Symfony/Component/Validator/Constraints/ValidValidator.php (hunk 1 above)
Comment (HeahDude) [inline src/Symfony/Component/Validator/Constraints/ValidValidator.php]: Looks clean to me like this.

**Review by HeahDude — APPROVED** [review]

**Review by nicolas-grekas — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Comment by Atiragram** [comment]
Hi,
What if value is null? i need @Assert\Valid only after and when @Assert\NotNull is valid, currently i get 500 error because of that empty value

**Comment by ogizanagi** [comment]
@Atiragram : This is already supposed to be fixed in https://github.com/symfony/symfony/pull/25297, available since 3.4.1. If you still encounter an issue, please open a new issue with a reproducer or a failing test case. Thanks!

**Pull request merged**